### PR TITLE
!left command now *always* sets end time with inferred hatch time prior to that...

### PR DESCRIPTION
**Pull request by Googer**
Almost completely redone time validation code, context sensitive for what time (meaning one of hatch, start, or end) is being set

### CODE REVIEW

**Comment by Googer** [comment]
The main part to really review and make sure it makes sense is the switch block(s) in the time type's validate() and parse() functions - those are what set the valid time ranges for any of the possible times being set (one of start, hatch, or end time).

File: data/settings.json [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
-  "default_raid_duration": 120,
-  "exclusive_raid_duration": 5760,
+  "standard_raid_incubate_duration": 60,
   "standard_raid_hatched_duration": 60,
+  "exclusive_raid_incubate_duration": 5760,
   "exclusive_raid_hatched_duration": 60,
   "start_clear_time": 10,
   "deletion_grace_time": 5,
-  "min_raid_hour": 6,
-  "max_raid_hour": 22,
```
Comment (ForcedUnknown) [inline data/settings.json:9]: Does this mean raids can be created whenever now?

**Comment by Googer** [comment]
Yes

File: data/settings.json (hunk 1 above)
Comment (Googer) [reply data/settings.json:9]: Yes

File: types/time.js [hunk 2]
```diff
@@ -18,12 +18,64 @@ class TimeType extends Commando.ArgumentType {
...
+				first_possible_time = now.clone().add(-hatched_duration, 'minutes');
+				max_duration = incubation_duration;
+				last_possible_time = raid_creation_time.clone().add(max_duration, 'minutes');
+				break;
+			}
+
+			case TimeParameter.END:
+				// End time - valid range is now through incubation plus hatch duration past creation time
+				first_possible_time = now;
+				max_duration = incubation_duration + hatched_duration;
```
Comment (ForcedUnknown) [inline types/time.js:75]: so the maximum end time / duration for a raid is now 2 hours? just making sure this is on purpose

**Comment by ForcedUnknown** [comment]
Looks good here, approved.

File: types/time.js (hunk 2 above)
Comment (Googer) [reply types/time.js:75]: It always was, this just makes how that duration comes about more explicit, clearer, and more configurable if Niantic changes things down the line. :)

**Pull request merged**